### PR TITLE
Add JavaScript direct upload support

### LINKED ISSUE #29: Add direct upload file_field using JavaScript
Through the DirectUploadsController we support generating blobs prior to direct upload, but we don't yet provide any convenience to complete the full story with the required JavaScript needed to actually perform the direct upload. We should.

Here's the broken JS I've been using to test that this approach actually works:

```javascript
window.fileUpload = function(file) {
  // FIXME: Block the form from submitting while we are doing the direct upload

  let data = new FormData()
  data.append("blob[filename]", file.name)
  data.append("blob[byte_size]", file.size)
  data.append("blob[content_type]", file.type)
  
  // FIXME: Compute proper md5 in base64
  data.append("blob[checksum]", "xxx23423")

  fetch("/rails/active_storage/direct_uploads", { method: "POST", body: data }).then(r => r.json()).then(function(directUploadDetails) {
    const reader = new FileReader()

    reader.onload = function(event) {
      fetch(directUploadDetails["url"], { method: "PUT", body: event.currentTarget.result}).then(function() {
        // FIXME: Set a hidden field on the form with the value of directUploadDetails["sgid"] 
        // to reference the uploaded file
        console.log(directUploadDetails)
      })
    }

    reader.readAsBinaryString(file)
  })
  
  // FIXME: Release the form for submission now that the files are done uploading
  return true
}
```

What I'd like to see is something like this:

```erb
<%= form_with(model: Message.new) do |form| %>
  <%= form.file_field :images, multiple: true, 'data-direct-upload': true %>
<% end %>
```

This should do the following:

1. When one or more files are selected, we call `/rails/active_storage/direct_uploads` for each with the details ala the broken JS above. This gives us a set of details for each.
2. We then disable the submit button on the form to prevent it from being sent before we're done upload.
3. Once each file is done uploading, we add a hidden field that provides the sgids and we reenable the submit button.

When this form is submitted, we'll verified the sgids in `#attach` and just associate them as though they were finally uploaded blobs.

### LINKED ISSUE #79: Get more informations on direct_uploads
Currently from the `DirectUploadsController` we return `{:url, :sgid}` which works fine with S3 and GCS since they sign all required fields in query_string. 
But, Azure require some informations to be passed in headers, currently that header is 
`x-ms-blob-type: "BlockBlob"`.

Can we return fatter response from the controller that contains `url, sgid, headers` and use him on the fontend somethig like this
```js
  fetch("/rails/active_storage/direct_uploads", { method: "POST", body: data }).then(r => r.json()).then(function(directUploadDetails) {
    const reader = new FileReader()

    reader.onload = function(event) {
      fetch(directUploadDetails["url"], { 
        method: "PUT",
        headers: directUploadDetails.headers, // something like this
        body: event.currentTarget.result}).then(function() {
          ...

      })
    }

    reader.readAsBinaryString(file)
  })
 
```

Also, I suppose there are other storage providers in the wild that are going to need something like this

### PR DESCRIPTION
**Pull request by javan**
🖥 → ☁️

Fixes #29
Probably fixes #79
Closes #35

### CODE REVIEW

File: app/helpers/active_storage/meta_helper.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module ActiveStorage::MetaHelper
+  def active_storage_meta_tags
+    safe_join [
+      tag.meta(name: 'direct-uploads-url', content: rails_direct_uploads_url)
+    ]
```
Comment (dhh) [inline app/helpers/active_storage/meta_helper.rb]: There's only one tag?

Comment (dhh) [inline app/helpers/active_storage/meta_helper.rb]: Also, I'm thinking that maybe we should just decorate the specific file input tag with the URL instead. Like we do with other data-url decorations. We'll wrap that into a direct_upload_file_tag anyway.

Comment (javan) [inline app/helpers/active_storage/meta_helper.rb]: Currently, yes. Leaves us open to adding more without introducing a name change [like `csrf_meta_tags`](https://github.com/rails/rails/blob/9c35bf2a6a27431c6aa283db781c19f61c5155be/actionview/lib/action_view/helpers/csrf_helper.rb#L31-L32).

Comment (javan) [inline app/helpers/active_storage/meta_helper.rb]: That's supported currently and will take precedence over the the meta tag value.

Comment (dhh) [inline app/helpers/active_storage/meta_helper.rb]: So the meta tag is just there to save repeating the URL on the input tags? That seems like a bit of an over-optimization to me. I'd rather just have it on the input tag alone, such that using this feature doesn't require any setup gymnastics.

**Comment by javan** [comment]
Demo / example using these direct upload events: https://gist.github.com/javan/5538692cb37a683db15792be8d05761e

**Comment by isaac** [comment]
@javan looks great! Is there a way to start the upload as soon as the user selects the file(s) rather than when the form is submitted?

**Comment by jeffreyguenther** [comment]
Quick question: Where's the source for the JS? I might be missing something, but I'm seeing a minified version in `app/assets/javascripts`.

**Comment by javan** [comment]
@isaac not currently, but you can implement that yourself by handling a file input's `change` event and creating [`DirectUpload`](https://github.com/rails/activestorage/blob/master/app/javascript/activestorage/direct_upload.js)s for its `File`s. See [`DirectUploadController`](https://github.com/rails/activestorage/blob/master/app/javascript/activestorage/direct_upload_controller.js) for example usage.

**Pull request merged**